fix: Localize remote editor

dcalhoun opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Fix localization of strings within the remote editor.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Close CMM-504. Close CMM-547. Localization worked in the local editor, but not in the remote editor.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
The remote editor relies upon accessing `@wordpress` modules via the
`window.wp` global. Most of these references are handled by
externalization performed by Vite. However, the `@wordpress/i18n` module
is a dependency of `@wordpress/api-fetch`, which we do not want to
externalize. So, the `import` of `@wordpress/i18n` is not correctly
externalized for the remote editor.

As a workaround, we check for and utilize the global if it is available.
Long term, we should align the local and remote editor architecture so
that they load in the same manner. At which point, we could remove the
conditional and always reference the global.

Additionally, I added a guard to referencing the Node.js `process` to avoid
intermittent server errors.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
1. Start the `make dev-server-remote` command.
2. Open the iOS demo app and set the `GUTENBERG_EDITOR_REMOTE_URL` environment
   variable to `http://localhost:5174/remote.html`.
3. Apply the below example diff to the iOS demo app, replacing the `<YOUR-SIMPLE-SITE-NAME>`
   and `<REDACTED>` placeholders with the actual site name and auth header.
4. Run the app.
5. Open the editor and verify that the UI strings are in Spanish.

<details><summary>Example Diff</summary>

```diff
diff --git a/ios/Demo-iOS/Gutenberg.xcodeproj/xcshareddata/xcschemes/Gutenberg.xcscheme b/ios/Demo-iOS/Gutenberg.xcodeproj/xcshareddata/xcschemes/Gutenberg.xcscheme
index 1a352df..b6fd3e5 100644
--- a/ios/Demo-iOS/Gutenberg.xcodeproj/xcshareddata/xcschemes/Gutenberg.xcscheme
+++ b/ios/Demo-iOS/Gutenberg.xcodeproj/xcshareddata/xcschemes/Gutenberg.xcscheme
@@ -54,12 +54,12 @@
          <EnvironmentVariable
             key = "GUTENBERG_EDITOR_URL"
             value = "http://localhost:5173/"
-            isEnabled = "NO">
+            isEnabled = "YES">
          </EnvironmentVariable>
          <EnvironmentVariable
             key = "GUTENBERG_EDITOR_REMOTE_URL"
             value = "http://localhost:5174/remote.html"
-            isEnabled = "NO">
+            isEnabled = "YES">
          </EnvironmentVariable>
       </EnvironmentVariables>
    </LaunchAction>
diff --git a/ios/Demo-iOS/Sources/EditorView.swift b/ios/Demo-iOS/Sources/EditorView.swift
index f1f3394..b71ce84 100644
--- a/ios/Demo-iOS/Sources/EditorView.swift
+++ b/ios/Demo-iOS/Sources/EditorView.swift
@@ -72,7 +72,13 @@ private struct _EditorView: UIViewControllerRepresentable {
     var editorURL: URL?
 
     func makeUIViewController(context: Context) -> EditorViewController {
-        let viewController = EditorViewController()
+        var config = EditorConfiguration()
+        config.locale = "es"
+        config.plugins = true
+        config.siteApiNamespace = ["sites/<YOUR-SIMPLE-SITE-NAME>.wordpress.com"]
+        config.siteApiRoot = "https://public-api.wordpress.com/"
+        config.authHeader = "Bearer <YOUR-AUTH-TOKEN>"
+        let viewController = EditorViewController(configuration: config)
         viewController.editorURL = editorURL
         if #available(iOS 16.4, *) {
             viewController.webView.isInspectable = true

```

</details>

### Accessibility Testing Instructions
<!-- How can you test the changes by using a keyboard/screen reader only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A, no visual changes. 

## Screenshots or screencast <!-- if applicable -->
N/A, no visual changes.

dcalhoun committed:
fix: Localize remote editor

The remote editor relies upon accessing `@wordpress` modules via the
`window.wp` global. Most of these references are handled by
externalization performed by Vite. However, the `@wordpress/i18n` module
is a dependency of `@wordpress/api-fetch`, which we do not want to
externalize. So, the `import` of `@wordpress/i18n` is not correctly
externalized for the remote editor.

As a workaround, we check for and utilize the global if it is available.
Long term, we should align the local and remote editor architecture so
that they load in the same manner. At which point, we could remove the
conditional and always reference the global.

dcalhoun committed:
fix: Mitigate cryptic ES module failures

Occasionally, the local editor fails to load entirely. Sometimes it
display no errors, other times it'll throw exceptions related to
references `process`. Guarding the `process` reference will help prevent
the `process` error in particular.

crazytonyli commented:
Does the `configuration.locale` only affect the hard-coded strings in the Gutenberg library? Does Gutenberg pass it to the site API requests, so that potentially the block name (for example, "Jetpack AI Assistant") can be localized? Also, why is the "Add a block" title and the tabs below don't use the right language?

<img src="https://github.com/user-attachments/assets/346af692-a206-457a-81e3-5d77d738bd87" width=400 />

dcalhoun commented:
> Does the `configuration.locale` only affect the hard-coded strings in the Gutenberg library? 

Yes, only strings within the Gutenberg project are localized at this time. We have issues for expanding l10n. 

> Does Gutenberg pass it to the site API requests, so that potentially the block name (for example, "Jetpack AI Assistant") can be localized? 

No, it does not. If this is possible, this is a great idea for localizing plugin-provided strings. 

> Also, why is the "Add a block" title and the tabs below don't use the right language?

This is surprising. I am unable to reproduce this in with either the Demo app or WP-iOS. 🤔 

Do you note any other missing strings—block toolbar menu items, block settings, etc? I presume your test is using the steps I provided? Do you experience the issue when testing with WP-iOS?

crazytonyli commented:
> Do you note any other missing strings—block toolbar menu items, block settings, etc?

The block is still in English, even though the block inserter is mostly Spanish.

<img src="https://github.com/user-attachments/assets/39349da6-a07c-4abc-b9ad-2206661da1c9" width=400 />

> I presume your test is using the steps I provided?

Yes, I double checked. I followed the steps. Except, my dev server address is at http://localhost:5173/remote.html, the 5174 port does not load for me.

> Do you experience the issue when testing with WP-iOS?

I didn't test WP-iOS with changes in this PR.

dcalhoun committed:
fix: Remote editor localization timeout

Dynamic importing the localization resulted in a timeout due to a cyclic
dependency with the remote entry file. Specifically, with vendor assets
such as Vite helper utilities.

dcalhoun committed:
refactor: Replace anonymous functions with function declarations

Improve readability of top-level logic.

dcalhoun committed:
refactor: Replace compat function with Vite import rewrite

Avoid boilerplate code by relying upon Vite to replace the import with a
global reference. The global `wp.i18n` reference is defined by the
scripts loaded from the site.

dcalhoun committed:
docs: Note importance of promises for dynamic editor loads

dcalhoun committed:
refactor: Local editor utilizes promises rather than async/await

Align implementation with the remote editor, which uses promises to
avoid timeouts from circular dependencies.

dcalhoun committed:
refactor: Rename remote editor functions for consistency

Align with local editor naming.

dcalhoun commented on vite.config.remote.js:
```diff
@@ -99,8 +67,8 @@ function wordPressExternals() {
 
 				if (
 					! externalDefinition ||
-					/@wordpress\/(api-fetch|i18n|url)/.test( id ) ||
-					/@wordpress\/(api-fetch|i18n|url)/.test( module )
```
`i18n` was removed to ensure the `localization` module correctly references the `window.wp.i18n` global for the remote editor.

dcalhoun commented:
> Yes, I double checked. I followed the steps. Except, my dev server address is at http://localhost:5173/remote.html, the 5174 port does not load for me.

Good to know what testing steps you following, thank you. The `5174` port is solely because I generally run both the local and remote editor servers side-by-side. 

> I didn't test WP-iOS with changes in this PR.

Right, I was requesting that you attempt testing in WP-iOS as I am unable to reproduce the issue in both GBK and WP-iOS currently. 

That said, I pushed additional refactors to further align the l10n logic across the local and remote editors, and hope that maybe this also resolves the issue you observe. Would you please test and let me know? Thanks! 🙇🏻‍♂️

crazytonyli commented:
From the WP-iOS app, I can see the translated "Add a block" text. However, I still see the English text in the Demo-iOS app
<img src="https://github.com/user-attachments/assets/bfc44639-892c-40f4-9955-35406cdbbfab" width=400 />

dcalhoun commented:
> From the WP-iOS app, I can see the translated "Add a block" text. However, I still see the English text in the Demo-iOS app 

@crazytonyli glad to hear the translated strings display as expected within the WP-iOS app. I'll merge this with that positive testing outcome.  

Given you and I experience different outcomes in the Demo app, it seems to be an environment configuration issue. Does the issue resolve if you run `make build`, which downloads the translation strings?

crazytonyli commented:
> Does the issue resolve if you run `make build`

I just tried it again, and it resolved the issue. However, I'm pretty sure I tried both ways (`make build` and `make dev-server-remote`) in previous tests too...

dcalhoun commented:
It makes sense to me that it might solve the problem (missing translations downloaded), but it is strange that the same approach did not work previously... 🤔
